### PR TITLE
bare minimum working object detection

### PR DESCRIPTION
**Pull request by dyc3**
Adds object detection when you take a picture. Also adds some util functions to make processing images easier.

image

### CODE REVIEW

**Comment by carpenterd777** [comment]
just a comment before I start a review proper, this is so cool - I'm glad we were able to get something working!

also my initial review is being done at school so I cannot test the app itself, so this is just a code review

File: package.json [hunk 1]
```diff
@@ -10,13 +10,21 @@
...
+		"expo-gl": "^10.4.2",
+		"expo-gl-cpp": "^10.4.1",
+		"expo-image-manipulator": "^9.2.2",
 		"expo-status-bar": "~1.0.4",
+		"jpeg-js": "^0.4.3",
 		"react": "16.13.1",
 		"react-dom": "16.13.1",
 		"react-native": "https://github.com/expo/react-native/archive/sdk-42.0.0.tar.gz",
+		"react-native-fs": "^2.18.0",
 		"react-native-web": "~0.13.12"
```
Comment (carpenterd777) [inline package.json:28]:
```suggestion
		"@tensorflow-models/coco-ssd": "^2.2.2",
		"@tensorflow/tfjs": "^3.9.0",
		"@tensorflow/tfjs-react-native": "^0.7.0",
		"expo": "~42.0.1",
		"expo-camera": "~11.2.2",
		"expo-status-bar": "~1.0.4",
		"jpeg-js": "^0.4.3",
		"react": "16.13.1",
		"react-dom": "16.13.1",
		"react-native": "https://github.com/expo/react-native/archive/sdk-42.0.0.tar.gz",
		"react-native-web": "~0.13.12"
```
some of these look like they are not being used, but feel free to correct me if I am wrong

File: components/CameraView.tsx [hunk 2]
```diff
@@ -1,16 +1,37 @@
...
+		const buffer = await FileSystem.readAsStringAsync(picture?.uri ?? "", {
+			encoding: FileSystem.EncodingType.Base64,
+		});
+		const imageTensor: tf.Tensor3D = base64ImageToTensor(buffer);
+
+		setStatus("Loading object detection model...");
+		const objectDetection = await cocoSsd.load();
+		setStatus("Detecting objects...");
+		const predictions = await objectDetection.detect(imageTensor);
+		setObjectDetectionPreidictions(predictions);
```
Comment (carpenterd777) [inline components/CameraView.tsx:33]: neat! looks simple but I'm sure there's a lot going on here haha

**Review by carpenterd777 — CHANGES_REQUESTED** [review]

File: util/image.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+	const buffer = new Uint8Array(width * height * 3);
+	let offset = 0; // offset into original data
+	for (let i = 0; i < buffer.length; i += 3) {
+		buffer[i] = data[offset];
+		buffer[i + 1] = data[offset + 1];
+		buffer[i + 2] = data[offset + 2];
+
+		offset += 4;
+	}
+	return tf.tensor3d(buffer, [height, width, 3]);
```
[the remark this answers is not in the document]
Comment (dyc3) [reply util/image.ts:33]: whoops, yup, meant to do that

File: package.json (hunk 1 above)
Comment (dyc3) [reply package.json:28]: I'll see if I can narrow it down a bit

Comment (dyc3) [reply package.json]: We are probably going to want `expo-image-manipulator` later for image resizing, but I've removed it for now. All the rest are necessary.

File: components/CameraView.tsx [hunk 4]
```diff
@@ -38,6 +61,8 @@ const CameraView = (): JSX.Element => {
 				type={Camera.Constants.Type.back}
 			/>
 			<PhotoButton onPress={onPressPhotoButton} color="#f4e1e6" />
+			<Text>{lastImageURI}</Text>
+			<Text>{status}</Text>
```
Comment (carpenterd777) [inline components/CameraView.tsx:65]:
```suggestion
			<Text>{lastImageURI}</Text>
			<Text>{status}</Text>
```

can we mark these as temporary somehow so we know to get rid of them later?

**Review by carpenterd777 — CHANGES_REQUESTED** [review]

File: components/CameraView.tsx (hunk 4 above)
Comment (dyc3) [reply components/CameraView.tsx:66]: added a comment

**Review by carpenterd777 — APPROVED** [review]

**Pull request merged**